dbms/src/Flash/Pipeline/Exec/PipelineExec.h

```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Operators/Operator.h>
+
+#include <memory>
+
+namespace DB
+{
+class PipelineExecutorStatus;
+
+// The executor for push model operator.
+// A pipeline will generate multiple pipeline_execs.
+// data flow: source --> transform --> .. --> transform --> sink
+class PipelineExec
+{
+public:
+    PipelineExec(
+        SourceOpPtr && source_op_,
+        TransformOps && transform_ops_,
+        SinkOpPtr && sink_op_)
+        : source_op(std::move(source_op_))
+        , transform_ops(std::move(transform_ops_))
+        , sink_op(std::move(sink_op_))
+    {}
+
+    OperatorStatus execute(PipelineExecutorStatus & exec_status);
+
+    OperatorStatus await(PipelineExecutorStatus & exec_status);
+
+    OperatorStatus spill(PipelineExecutorStatus & exec_status);
+
+private:
+    OperatorStatus executeImpl(PipelineExecutorStatus & exec_status);
+
+    OperatorStatus awaitImpl(PipelineExecutorStatus & exec_status);
+
+    OperatorStatus spillImpl(PipelineExecutorStatus & exec_status);
+
+    OperatorStatus fetchBlock(
+        Block & block,
+        size_t & start_transform_op_index,
+        PipelineExecutorStatus & exec_status);
+
+    // It will only be called by `executeImpl` and `awaitImpl`.
+    // Because the spilling state will only be transferred from spilling status and running status.
+    template <typename Op>
+    OperatorStatus prepareSpillOpIfNeed(OperatorStatus op_status, const Op & op)
+    {
+        assert(op);
+        if (op_status == OperatorStatus::SPILLING)
+        {
+            assert(!ready_spill_op);
+            ready_spill_op.emplace(op.get());
+        }
+        return op_status;
```

Comment:
Maybe we shouldn't let this method pass-through the passed `op_status` as a return value. How about just `void` it and let callers return the `op_status` instead?

Reply:
ok

Reply:
updated.